ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/linear_optimisation/algorithms/fillers/RaUsageLimitsFiller.java

```diff
@@ -0,0 +1,154 @@
+ * maximum number of RAs per TSO (max-ra-per-tso), and maximum number of PSTs per TSO (max-pst-per-tso).
+ * <!> Introduces binary variables to define if an RA is used. </!>
+ *
+ * @author Peter Mitri {@literal <peter.mitri at rte-france.com>}
+ */
+public class RaUsageLimitsFiller implements ProblemFiller {
+    private final Set<RangeAction<?>> rangeActions;
+    private final RangeActionResult prePerimeterRangeActionResult;
+    private final Integer maxRa;
+    private final Integer maxTso;
+    private final Set<String> maxTsoExclusions;
+    private final Map<String, Integer> maxPstPerTso;
+    private final Map<String, Integer> maxRaPerTso;
+
+    public RaUsageLimitsFiller(Set<RangeAction<?>> rangeActions,
+                               RangeActionResult prePerimeterRangeActionResult,
+                               Integer maxRa,
+                               Integer maxTso,
+                               Set<String> maxTsoExclusions,
+                               Map<String, Integer> maxPstPerTso,
+                               Map<String, Integer> maxRaPerTso) {
+        this.rangeActions = rangeActions;
+        this.prePerimeterRangeActionResult = prePerimeterRangeActionResult;
+        this.maxRa = maxRa;
+        this.maxTso = maxTso;
+        this.maxTsoExclusions = maxTsoExclusions != null ? maxTsoExclusions : new HashSet<>();
+        this.maxPstPerTso = maxPstPerTso;
+        this.maxRaPerTso = maxRaPerTso;
+    }
+
+    @Override
+    public void fill(LinearProblem linearProblem, FlowResult flowResult, SensitivityResult sensitivityResult) {
+        if ((maxRa == null || maxRa >= rangeActions.size())
+            && (maxTso == null || maxTso >= rangeActions.stream().map(RemedialAction::getOperator).distinct().filter(tso -> !maxTsoExclusions.contains(tso)).count())
+            && (maxPstPerTso == null || maxPstPerTso.isEmpty())
+            && (maxRaPerTso == null || maxRaPerTso.isEmpty())) {
+            return;
+        }
+        rangeActions.forEach(ra -> buildIsVariationVariableAndConstraints(linearProblem, ra));
+        addMaxRaConstraint(linearProblem);
+        addMaxTsoConstraint(linearProblem);
+        addMaxRaPerTsoConstraint(linearProblem);
+        addMaxPstPerTsoConstraint(linearProblem);
+    }
+
+    private void buildIsVariationVariableAndConstraints(LinearProblem linearProblem, RangeAction<?> rangeAction) {
+        MPVariable isVariationVariable = linearProblem.addRangeActionVariationBinary(rangeAction);
+        double initialSetpoint = prePerimeterRangeActionResult.getOptimizedSetPoint(rangeAction);
+        MPVariable setpointVariable = linearProblem.getRangeActionSetpointVariable(rangeAction);
+        if (setpointVariable == null) {
+            throw new FaraoException(format("Range action setpoint variable for %s has not been defined yet.", rangeAction.getId()));
+        }
+
+        // range action setpoint <= intial setpoint + isVariationVariable * (max setpoint - initial setpoint)
```

Comment:
Moreover, I'm also thinking that, in MIP mode, it could be more interesting for the optimization to have some 'slack' on the definition of the initial_setpoint.

Indeed, this initial_setpoint might be not reachable, due to the fact that the conversion tap->angle is not perfect in MIP mode.

So, in the spirit, having something like:

range action setpoint <= intial_setpoint_safeminus + isVariationVariable * (max_setpoint_plus_epsilon - intial_setpoint_safeminus)

with intial_setpoint_safeminus = initial_setpoint - 1/4 (Delta_angle_between_taps)


And something similar with the other constraints with :
with intial_setpoint_safeminus = initial_setpoint + 1/4 (Delta_angle_between_taps)

(only in MIP, and not in CONTINUOUS)